test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -923,12 +944,8 @@ class test_ucp_sockaddr : public ucp_test {
             uct_iface_h uct_iface = ep->uct_eps[lane]->iface;
             auto res = sender_uct_ops.emplace(uct_iface, uct_iface->ops);
             if (res.second) {
-                uct_iface->ops.ep_flush =
-                        reinterpret_cast<uct_ep_flush_func_t>(
-                                ucs_empty_function_return_no_resource);
-                uct_iface->ops.ep_pending_add =
-                        reinterpret_cast<uct_ep_pending_add_func_t>(
-                                ucs_empty_function_return_busy);
+                uct_iface->ops.ep_flush       = ep_flush;
+                uct_iface->ops.ep_pending_add = ep_pending_add;
```

Comment:
what if we replaced ep_flush (and not changed ops.ep_pending_add)?
for example, could return UCS_ERR_NO_RESOURCE or UCS_INPROGRESS from the flush, to prevent completion of discard operation

Reply:
but then we need to save the completion and then complete it with the status of `uct_ep_flush()` completion which should be called from gtest.

Reply:
what if returning UCS_ERR_NO_RESOURCE?

Reply:
it seems that in this PR, the same loop can happen when the callback is  ucp_worker_discard_uct_ep_flush_comp ?

Reply:
> what if returning UCS_ERR_NO_RESOURCE?

currently, this function returns `UCS_ERR_NO_RESOURCE`.

> it seems that in this PR, the same loop can happen when the callback is ucp_worker_discard_uct_ep_flush_comp ?

yes, it will do `uct_ep_flish` and `uct_ep_pending_add`, but we register UCT worker progress function to do new round, so it leaves `uct_worker_progress`.

Reply:
> currently, this function returns UCS_ERR_NO_RESOURCE.
So maybe we don't need to overwrite pending_add?

Reply:
> > currently, this function returns UCS_ERR_NO_RESOURCE.

> So maybe we don't need to overwrite pending_add?

Yes, it should be ok to rely on UCT here.
Done

Reply:
testing discovered the issue and it's correct - if `pending_cb` returns `UCS_ERR_NO_RESOUCE` then the element isn't removed from the pending queue and it means that no resources to do an operation.
```
[swx-rdmz-ucx-new-02:22712:0:22712]      tcp_ep.c:959  Assertion `ucs_queue_is_empty(&ep->pending_q)' failed

/scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/uct/tcp/tcp_ep.c: [ uct_tcp_ep_pending_queue_dispatch() ]
      ...
      956     uct_pending_queue_dispatch(priv, &ep->pending_q,
      957                                uct_tcp_ep_ctx_buf_empty(&ep->tx));
      958     if (uct_tcp_ep_ctx_buf_empty(&ep->tx)) {
==>   959         ucs_assert(ucs_queue_is_empty(&ep->pending_q));
      960         uct_tcp_ep_mod_events(ep, 0, UCS_EVENT_SET_EVWRITE);
      961     }
      962 }

==== backtrace (tid:  22712) ====
 0 0x0000000000024885 uct_tcp_ep_pending_queue_dispatch()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/uct/tcp/tcp_ep.c:959
 1 0x000000000002d8a6 uct_tcp_cm_change_conn_state()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/uct/tcp/tcp_cm.c:58
 2 0x000000000002e871 uct_tcp_cm_handle_conn_req()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/uct/tcp/tcp_cm.c:613
 3 0x000000000002e871 uct_tcp_cm_handle_conn_pkt()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/uct/tcp/tcp_cm.c:692
 4 0x000000000002661a uct_tcp_ep_progress_am_rx()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/uct/tcp/tcp_ep.c:1441
 5 0x000000000002661a uct_tcp_ep_progress_data_rx()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/uct/tcp/tcp_ep.c:1515
 6 0x000000000002ae94 uct_tcp_iface_handle_events()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/uct/tcp/tcp_iface.c:310
 7 0x0000000000073b01 ucs_event_set_wait()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucs/sys/event_set.c:215
 8 0x000000000002ada4 uct_tcp_iface_progress()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/uct/tcp/tcp_iface.c:327
 9 0x00000000000575a2 ucs_callbackq_dispatch()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucs/datastruct/callbackq.h:211
10 0x00000000000575a2 uct_worker_progress()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/uct/api/uct.h:2600
11 0x00000000000575a2 ucp_worker_progress()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucp/core/ucp_worker.c:2657
12 0x0000000000a152cc ucp_test_base::entity::progress()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/ucp_test.cc:1047
13 0x0000000000a154ff ucp_test::progress()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/ucp_test.cc:168
14 0x0000000000a154ff ucp_test::request_progress()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/ucp_test.cc:266
15 0x0000000000a0cd3f test_ucp_sockaddr::do_force_close_during_rndv()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_sockaddr.cc:949
16 0x0000000000a0cd3f ~vector()  /usr/include/c++/4.8.2/bits/stl_vector.h:416
17 0x0000000000a0cd3f test_ucp_sockaddr::do_force_close_during_rndv()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_sockaddr.cc:949
18 0x00000000005f6cc6 ucs::test_base::run()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/test.cc:356
19 0x00000000005f6cc6 ucs::test_base::TestBodyProxy()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/test.cc:382
20 0x0000000000cb3a49 HandleSehExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/googletest/gtest.cc:2433
21 0x0000000000caa879 testing::Test::Run()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/googletest/gtest.cc:2509
22 0x0000000000caa9a1 testing::TestInfo::Run()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/googletest/gtest.cc:2687
23 0x0000000000caaa65 testing::TestSuite::Run()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/googletest/gtest.cc:2819
24 0x0000000000cab58a testing::internal::UnitTestImpl::RunAllTests()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/googletest/gtest.cc:5350
25 0x0000000000cab741 HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/googletest/gtest.cc:2433
26 0x000000000057d4a5 RUN_ALL_TESTS()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/googletest/gtest.h:2473
27 0x000000000057d4a5 main()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/main.cc:118
28 0x00000000000223d5 __libc_start_main()  ???:0
29 0x00000000005da0b4 _start()  ???:0
=================================
```
So, we should replace `pending_add()` by our own implementation to not add the callback to the pending and just always return `UCS_ERR_BUSY`.
@yosefe wdyt?

Reply:
per offline discussion, will try to check if discarding hash is empty when request is completed, instead of delaying discard and replacing function pointers

Reply:
>will try to check if discarding hash is empty when request is completed, instead of delaying discard and replacing function pointers

it works for `rcx/test_ucp_sockaddr.force_close_during_rndv/*` tests, but doesn't work for `rcx/test_ucp_sockaddr.fail_and_force_close_during_rndv/*` tests, because it receives ATS:
```
/hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/ucp/test_ucp_sockaddr.cc: [ test_ucp_sockaddr::scomplete_reset_data_cbx() ]
      ...
      302         scomplete_reset_data_arg_t *arg =
      303                 reinterpret_cast<scomplete_reset_data_arg_t*>(user_data);
      304 #if UCS_ENABLE_ASSERT
==>   305         ucs_assert(arg->ep->refcounts.discard == 0);
      306         ASSERT_EQ(0u, arg->ep->refcounts.discard);
      307 #endif
      308         arg->send_buffer->pattern_fill(0, arg->send_buffer->size());

==== backtrace (tid: 195328) ====
 0 0x0000000000a1d206 test_ucp_sockaddr::scomplete_reset_data_cbx()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/ucp/test_ucp_sockaddr.cc:305
 1 0x00000000000fc25d ucp_request_complete_send()  /hpc/mtr_scrap/users/dmitrygla/ucx/src/ucp/core/ucp_request.inl:234
 2 0x00000000000fc25d ucp_request_complete_and_dereg_send()  /hpc/mtr_scrap/users/dmitrygla/ucx/src/ucp/core/ucp_request.inl:1154
 3 0x00000000000fc25d ucp_rndv_ats_handler_inner()  /hpc/mtr_scrap/users/dmitrygla/ucx/src/ucp/rndv/rndv.c:1619
 4 0x00000000000fc25d ucp_rndv_ats_handler()  /hpc/mtr_scrap/users/dmitrygla/ucx/src/ucp/rndv/rndv.c:1598
 5 0x00000000000a1643 uct_iface_invoke_am()  /hpc/mtr_scrap/users/dmitrygla/ucx/src/uct/base/uct_iface.h:857
 6 0x00000000000a4ea8 uct_rc_mlx5_iface_common_am_handler()  /hpc/mtr_scrap/users/dmitrygla/ucx/src/uct/ib/rc/accel/rc_mlx5.inl:422
 7 0x00000000000a4ea8 uct_rc_mlx5_iface_common_poll_rx()  /hpc/mtr_scrap/users/dmitrygla/ucx/src/uct/ib/rc/accel/rc_mlx5.inl:1455
 8 0x00000000000a4ea8 uct_rc_mlx5_iface_progress()  /hpc/mtr_scrap/users/dmitrygla/ucx/src/uct/ib/rc/accel/rc_mlx5_iface.c:169
 9 0x00000000000a4ea8 uct_rc_mlx5_iface_progress_cyclic()  /hpc/mtr_scrap/users/dmitrygla/ucx/src/uct/ib/rc/accel/rc_mlx5_iface.c:179
10 0x0000000000060d66 ucs_callbackq_dispatch()  /hpc/mtr_scrap/users/dmitrygla/ucx/src/ucs/datastruct/callbackq.h:211
11 0x000000000006dc84 uct_worker_progress()  /hpc/mtr_scrap/users/dmitrygla/ucx/src/uct/api/uct.h:2600
12 0x0000000000a89fc4 ucp_test_base::entity::progress()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/ucp/ucp_test.cc:1047
13 0x0000000000a84af2 ucp_test::progress()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/ucp/ucp_test.cc:168
14 0x0000000000a84ef7 ucp_test::check_events()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/ucp/ucp_test.cc:249
15 0x0000000000a851bd ucp_test::request_process()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/ucp/ucp_test.cc:291
16 0x0000000000a85301 ucp_test::request_wait()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/ucp/ucp_test.cc:311
17 0x0000000000a85340 ucp_test::requests_wait()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/ucp/ucp_test.cc:320
18 0x0000000000a2256a test_ucp_sockaddr::do_force_close_during_rndv()  /hpc/mtr_scrap/users/dmitrygla/ucx/test/gtest/ucp/test_ucp_sockaddr.cc:996
...
```

Also, I'm not sure that it is a good idea to rely on `ucp_ep_t::refcounts::discard`, because we're losing the main idea of the test that the user could read data that was already invalidated.